clang/labs/lab1/alexseev_danila/deprecatedWarning.cpp

```diff
@@ -0,0 +1,57 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+
+using namespace clang;
+
+class DepFuncVisitor : public RecursiveASTVisitor<DepFuncVisitor> {
+private:
+  ASTContext *Context;
+
+public:
+  explicit DepFuncVisitor(ASTContext *Context) : Context(Context) {}
+
+  bool VisitFunctionDecl(FunctionDecl *Func) {
+    if (Func->getNameInfo().getAsString().find("deprecated") !=
+        std::string::npos) {
+      DiagnosticsEngine &Diags = Context->getDiagnostics();
+      size_t CustomDiagID =
+          Diags.getCustomDiagID(DiagnosticsEngine::Warning,
+                                "Function contains 'deprecated' in its name");
+      Diags.Report(Func->getLocation(), CustomDiagID)
+          << Func->getNameInfo().getAsString();
+    }
+    return true;
+  }
+};
+
+class DepFuncConsumer : public ASTConsumer {
+private:
+  CompilerInstance &Instance;
+
+public:
+  explicit DepFuncConsumer(CompilerInstance &CI) : Instance(CI) {}
+
+  void HandleTranslationUnit(ASTContext &Context) override {
+    DepFuncVisitor Visitor(&Instance.getASTContext());
+    Visitor.TraverseDecl(Context.getTranslationUnitDecl());
+  }
+};
+
+class DepFuncPlugin : public PluginASTAction {
+protected:
+  std::unique_ptr<ASTConsumer>
+  CreateASTConsumer(CompilerInstance &Compiler,
+                    llvm::StringRef InFile) override {
+    return std::make_unique<DepFuncConsumer>(Compiler);
+  }
+
+  bool ParseArgs(const CompilerInstance &Compiler,
+                 const std::vector<std::string> &Args) override {
+    return true;

```

Comment:
Please add an option specifying a function name to exclude from check.

Reply:
I added the '-exclude' option which takes an argument (function name). Example: '-exclude=deprecatedFunc'. When this option is used, a function named 'deprecatedFunc' will not generate a warning. I also added a check of the option’s operation to the test.